actions/image_partition_action.go

```diff
@@ -438,16 +446,18 @@ func (i ImagePartitionAction) Run(context *debos.DebosContext) error {
 	defer i.triggerDeviceNodes(context)
 	defer imageFD.Close()
 
-	err = syscall.Flock(int(imageFD.Fd()), syscall.LOCK_EX)
-	if err != nil {
-		return err
-	}
+	lockImageExclusively(imageFD)
 
 	command := []string{"parted", "-s", context.Image, "mklabel", i.PartitionType}
 	if len(i.GptGap) > 0 {
 		command = append(command, i.GptGap)
 	}
+	/* Unlock image for parted call; then lock afterwards.
+	 * See https://github.com/go-debos/debos/issues/344#issuecomment-1275835432
+	 */
+	lockImageShared(imageFD)
 	err = debos.Command{}.Run("parted", command...)
+	lockImageExclusively(imageFD)
```

Comment:
2 things; first this code is copied 3 times; almost like it should be a runParted function instead :)

But more importantly i'm not sure if this is the right thing, the whole reason for exclusive locking was to avoid udev trigger the kernel to rescan the partition causing block devices to temporarily disappear. So it's re-opening the window for the old race condition again unfortunately

Reply:
@sjoerdsimons do you have a suggestion on a way I can move this forward?

Reply:
I think the first thing to do here is to make sure we have a testcase again for the old error case; What is happening is quite well described on the commit and the original issue, but i've been failing to reproduce it last night with the flock introducing commit reverted which is quite annoying.

Once we can reproduce it a potential way to fix this without temporarily dropping flock's (which makes taking it somewhat  useless i think, it just potentialy makes the race window samller) could be to only take the lock *after* all partitioning is done and then have debos validate all partition devices have (re)appeared  before continuing with the formatting pass.  

As this would then block all further udev processing while waiting for the kernel to do all the reprobing and recreation of devices. The one thing to verify for that though would be to ensure the `BLKRRPART` ioctl *only* returns *after* dropping the previous set of partition devices which would mean it's safe to assume that when they reappear the operation has finished. 

Fwiw this area is nasty in that it's a _very_ small race window (which is why it took so long previous to reproduce and work out what was going on); So i want to be really sure we don't re-introduce races as that will come back and bite us for sure :)